Tests/SourceKitLSPTests/RewriteSourceKitPlaceholdersTests.swift

```diff
@@ -0,0 +1,62 @@
+import SKTestSupport
+@testable import SourceKitLSP
+import XCTest
+
+final class RewriteSourceKitPlaceholdersTests : XCTestCase {
+  func testClientDoesNotSupportSnippets() {
+    let input = "foo(bar: \(placeholder: "T##Int##Int#"))"
+    let rewritten = rewriteSourceKitPlaceholders(in: input, clientSupportsSnippets: false)
+
+    XCTAssertEqual(rewritten, "foo(bar: )")
+  }
+
+  func testInputWithoutPlaceholders() {
+    let input = "foo()"
+    let rewritten = rewriteSourceKitPlaceholders(in: input, clientSupportsSnippets: true)
+
+    XCTAssertEqual(rewritten, "foo()")
+  }
+
+  func testPlaceholderWithType() {
+    let input = "foo(bar: \(placeholder: "T##bar##Int"))"
+    let rewritten = rewriteSourceKitPlaceholders(in: input, clientSupportsSnippets: true)
+
+    XCTAssertEqual(rewritten, "foo(bar: ${1:Int})")
+  }
+
+  func testMultiplePlaceholders() {
+    let input = "foo(bar: \(placeholder: "T##Int##Int"), baz: \(placeholder: "T##String##String"), quux: \(placeholder: "T##String##String"))"
+    let rewritten = rewriteSourceKitPlaceholders(in: input, clientSupportsSnippets: true)
+
+    XCTAssertEqual(rewritten, "foo(bar: ${1:Int}, baz: ${2:String}, quux: ${3:String})")
+  }
+
+  func testClosurePlaceholderReturnType() {
+    let input = "foo(bar: \(placeholder: "{ \(placeholder: "T##Int##Int") }"))"
+    let rewritten = rewriteSourceKitPlaceholders(in: input, clientSupportsSnippets: true)
+
+    XCTAssertEqual(rewritten, "foo(bar: ${1:\\{ ${2:Int} \\}})")
+  }
+
+  func testClosurePlaceholderArgumentType() {
+    let input = "foo(bar: \(placeholder: "{ \(placeholder: "T##Int##Int") in \(placeholder: "T##Void##Void") }"))"
+    let rewritten = rewriteSourceKitPlaceholders(in: input, clientSupportsSnippets: true)
+
+    XCTAssertEqual(rewritten, "foo(bar: ${1:\\{ ${2:Int} in ${3:Void} \\}})")
+  }
+
+  func testMultipleClosurePlaceholders() {
+    let input = "foo(\(placeholder: "{ \(placeholder: "T##Int##Int") }"), baz: \(placeholder: "{ \(placeholder: "Int") in \(placeholder: "T##Bool##Bool") }"))"
+    let rewritten = rewriteSourceKitPlaceholders(in: input, clientSupportsSnippets: true)
+
+    XCTAssertEqual(rewritten, "foo(${1:\\{ ${2:Int} \\}}, baz: ${3:\\{ ${4:Int} in ${5:Bool} \\}})")
+  }
+}
+
+private let placeholderStart = "<#"
+private let placeholderEnd = "#>"
+private extension DefaultStringInterpolation {
+  mutating func appendInterpolation(placeholder string: String) {
+    self.appendLiteral(placeholderStart + string + placeholderEnd)
+  }
+}
```

Comment:
I don’t think I’m a fan of extending string interpolation in this way. I would prefer to spell out the placeholders in the tests as `<#…#>`. I know that it’s a little annoying when editing the file in Xcode but I think overall I’ve found that it’s easiest to reason about these sorts of tests if you can view the input text as it will returned from sourcekitd, even if that means opening the file in a non-Xcode editor at times.